spec/abbreviato/abbreviato_spec.rb

```diff
@@ -0,0 +1,373 @@
+      with: { max_length: 62 },
+      source: "<really_a_very_long_tag_name>text</really_a_very_long_tag_name>",
+      expected: "<really_a_very_long_tag_name>...</really_a_very_long_tag_name>"
+    it_does_not_truncate "does not allow closing tags to get added without opening tags",
+      with: { max_length: 63 },
+      source: "<really_a_very_long_tag_name>text</really_a_very_long_tag_name>",
+      expected: "<really_a_very_long_tag_name>text</really_a_very_long_tag_name>"
+  end
+
+  describe "edge-cases:" do
+    it_truncates "discard a single element within a longer element and use the following html (assuming it will fit)",
+      with: { max_length: 9 },
+      source: "<span><input/></span><p>.</p>",
+      expected: "<p>.</p>"
+    it_truncates "discard a single element within a longer element",
+      with: { max_length: 10 },
+      source: "<span><p></p></span><h1>.</h1>",
+      expected: "<h1>.</h1>"
+    it_truncates "some semi-random elements",
+      with: { max_length: 10 },
+      source: "<span><p><br/><br/><br/><p><br/></span><h1></h2>.</h3><h4><br/>",
+      expected: "<h1>.</h1>"
+    it_does_not_truncate "junk, including various html chars",
+      with: { max_length: 10 },
+      source: "<<< /  > < 0)(*&^*&^%${#}><? < /",
+      expected: ""
+    it_truncates "outer tags fit exactly",
+      with: { max_length: 13 },
+      source: "<span><p></p></span>",
+      expected: "<span></span>"
+    it_truncates "outer tags and opening inner tag fit exactly",
+      with: { max_length: 16 },
+      source: "<span><p></p></span>",
+      expected: "<span></span>"
+    it_truncates "void tags which do not fit",
+      with: { max_length: 5 },
+      source: "<command>",
+      expected: ""
+    it_truncates "void tags within outer tags which do not fit",
+      with: { max_length: 15 },
+      source: "<p><command/></p>",
+      expected: "<p></p>"
+    it_does_not_truncate "void tags within outer tags which fit",
+      with: { max_length: 17 },
+      source: "<p><command/></p>",
+      expected: "<p><command/></p>"
+    it_truncates "outer tags which fit perfectly within inner content",
+      with: { max_length: 7 },
+      source: "<p>hello</p>",
+      expected: "<p></p>"
+  end
+
+  describe "void tags" do
+    TruncatedSaxDocument::VOID_TAGS.each do |tag|
+      it_does_not_truncate "void tag: #{tag} doesn't get closing element added",
+        with: { max_length: 100 },
+        source: "<#{tag}/>",
+        expected: "<#{tag}/>"
+
+      it_does_not_truncate "void tag: #{tag} has closing element stripped",
+        with: { max_length: 100 },
+        source: "<#{tag}></#{tag}>",
+        expected: "<#{tag}/>"
+    end
+  end
+
+  describe "fragment mode" do
+    it_truncates "doesn't add `head` and `body` tags",
+      with: { max_length: 15 },
+      source: "<p>hello there</p>",
+      expected: "<p>hello...</p>"
+  end
+
+  describe "document mode" do
```

Comment:
Are test cases needed to handle other charsets with encoded characters? Like ISO-8859-1: Å, å, Ä, ä, Ö, ö etc?